### PR TITLE
BUG: df.sort_values() not respecting na_position with categoricals #22556

### LINKED ISSUE #22556: df.sort_values() not respecting na_position with categoricals
#### Problem description
`DataFrame.sort_values()` appears not to respect the `na_position` parameter when sorting by a categorical series:

```python
>>> import pandas as pd
>>> c = pd.Categorical(['A', np.nan, 'B'], categories=['A','B'], ordered=True)
>>> df = pd.DataFrame({'c': c})
>>> df.sort_values(by='c', na_position='first')
     c
1  NaN
0    A
2    B
>>> df.sort_values(by='c', na_position='last')
     c
1  NaN
0    A
2    B
```
Unexpectedly, the NaNs always come first regardless of `na_position`.

#### Additional information
`Series.sort_values()` works as expected:
```python
>>> c.sort_values(na_position='first')
[NaN, A, B]
Categories (2, object): [A < B]
>>> c.sort_values(na_position='last')
[A, B, NaN]
Categories (2, object): [A < B]
```

Strangely, `df.sort_values()` does seem to respect `na_position` if you sort by more than one column (even the same column):
```python
>>> df.sort_values(by=['c','c'], na_position='first')
     c
1  NaN
0    A
2    B
>>> df.sort_values(by=['c','c'], na_position='last')
     c
0    A
2    B
1  NaN
```

#### Output of ``pd.show_versions()``

<details>

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.4
pytest: None
pip: 10.0.1
setuptools: 40.0.0
Cython: None
numpy: 1.15.0
scipy: None
pyarrow: None
xarray: None
IPython: 6.5.0
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
```
</details>

### PR DESCRIPTION
**Pull request by staftermath**
- [x] closes #22556 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/sorting.py [hunk 1]
```diff
@@ -241,7 +241,23 @@ def nargsort(items, kind='quicksort', ascending=True, na_position='last'):
 
     # specially handle Categorical
     if is_categorical_dtype(items):
-        return items.argsort(ascending=ascending, kind=kind)
+        mask = isna(items)
+        null_idx = np.where(mask)[0]
+        sorted_idx = items.argsort(ascending=ascending, kind=kind)
```
Comment (jreback) [inline pandas/core/sorting.py]: do we not have a .sort_values on Categorical that accepts na_position already?

if we don’t all of this code should live there anyhow

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/sorting.py (hunk 1 above)
Comment (staftermath) [reply pandas/core/sorting.py]: the way I see it, `df.sort_values` is not calling `sort_values` method from column objects, it's calling `nargsort`. Since `pd.Series` also doesn't have `nargsort`, I feel it may fit better in here rather than `Categorical.nargsort`

**Comment by staftermath** [comment]
**ci/circleci: py35_ascii**  failing related to this?: https://github.com/pandas-dev/pandas/issues/21763

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/sorting.py [hunk 2]
```diff
@@ -241,7 +241,19 @@ def nargsort(items, kind='quicksort', ascending=True, na_position='last'):
 
     # specially handle Categorical
     if is_categorical_dtype(items):
-        return items.argsort(ascending=ascending, kind=kind)
+        if na_position not in ['first', 'last']:
+            raise ValueError('invalid na_position: {!r}'.format(na_position))
+        mask = isna(items)
+        cnt_null = mask.sum()
+        sorted_idx = items.argsort(ascending=ascending, kind=kind)
+        if ascending:
```
[the remark this answers is not in the document]
Comment (staftermath) [reply pandas/core/sorting.py]: good point. not sure how to do condition check and negate `cnt_null` only once. updated to keep the `np.roll` only when needed and took return outside of condition.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/sorting.py [hunk 3]
```diff
@@ -241,7 +241,18 @@ def nargsort(items, kind='quicksort', ascending=True, na_position='last'):
...
     # specially handle Categorical
     if is_categorical_dtype(items):
-        return items.argsort(ascending=ascending, kind=kind)
+        if na_position not in ['first', 'last']:
+            raise ValueError('invalid na_position: {!r}'.format(na_position))
+        mask = isna(items)
+        cnt_null = mask.sum()
+        sorted_idx = items.argsort(ascending=ascending, kind=kind)
+        if ascending and na_position == 'last':
+            # NaN is coded as -1 and is listed in front after sorting
```
[the remark this answers is not in the document]
Comment (staftermath) [reply pandas/core/sorting.py]: great suggestion. updated.

File: pandas/tests/frame/test_sorting.py [hunk 4]
```diff
@@ -598,3 +598,67 @@ def test_sort_index_intervalindex(self):
...
+    def test_sort_index_na_position_with_categories(self):
+        # GH 22556
+        # Positioning missing value properly when column is Categorical.
+        categories = ['A', 'B', 'C']
+        list_of_nans = [np.nan, np.nan]
+        df = pd.DataFrame({
+            'c': pd.Categorical(['A', np.nan, 'B', np.nan, 'C'],
+                                categories=categories,
+                                ordered=True)})
+        result_ascending_na_first = df.sort_values(by='c',
```
Comment (jreback) [inline pandas/tests/frame/test_sorting.py]: rather than naming result / expected like this, can you just name them result & expected, and put a comment before the test case describing; its easier to read.

If you *can* easily parameterize things would be great as well (though may not be so easy).

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. pls rebase on master and ping on green.

File: pandas/tests/frame/test_sorting.py (hunk 4 above)
Comment (staftermath) [reply pandas/tests/frame/test_sorting.py]: @jreback thanks for all the suggestions, I have made the changes except for the parameterization (couldn't figure out a good way to do that.. )

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@WillAyd over to you

File: pandas/tests/frame/test_sorting.py [hunk 5]
```diff
@@ -598,3 +598,71 @@ def test_sort_index_intervalindex(self):
...
+        # sort ascending with na first
+        result = df.sort_values(by='c',
+                                ascending=True,
+                                na_position='first')
+        expected = DataFrame({
+            'c': Categorical(list_of_nans + categories,
+                             categories=categories,
+                             ordered=True)}, index=[1, 3, 0, 2, 4])
+
+        assert_frame_equal(result,
```
Comment (WillAyd) [inline pandas/tests/frame/test_sorting.py]: Readability nit but can you make this statement one line (applicable in a few instances below)

File: pandas/tests/frame/test_sorting.py [hunk 6]
```diff
@@ -598,3 +598,71 @@ def test_sort_index_intervalindex(self):
             closed='right')
         result = result.columns.levels[1].categories
         tm.assert_index_equal(result, expected)
+
+    def test_sort_index_na_position_with_categories(self):
```
Comment (WillAyd) [inline pandas/tests/frame/test_sorting.py]: Can you parametrize this test? Believe you can use `values`, `na_position` and `index` as parametrized arguments

File: pandas/core/sorting.py [hunk 7]
```diff
@@ -241,7 +241,19 @@ def nargsort(items, kind='quicksort', ascending=True, na_position='last'):
 
     # specially handle Categorical
     if is_categorical_dtype(items):
-        return items.argsort(ascending=ascending, kind=kind)
+        if na_position not in ['first', 'last']:
```
Comment (WillAyd) [inline pandas/core/sorting.py]: Minor but can change the list to a set here for inclusion testing

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_sorting.py (hunk 6 above)
Comment (staftermath) [reply pandas/tests/frame/test_sorting.py]: @WillAyd parametrized a bit. I am pretty new at this. Thanks for all the suggestions!

File: pandas/tests/frame/test_sorting.py [hunk 8]
```diff
@@ -598,3 +598,86 @@ def test_sort_index_intervalindex(self):
...
         result = result.columns.levels[1].categories
         tm.assert_index_equal(result, expected)
+
+    @pytest.mark.parametrize("categories", [['A', 'B', 'C']])
+    @pytest.mark.parametrize("category_indices", [[0, 2, 4]])
+    @pytest.mark.parametrize("list_of_nans", [[np.nan, np.nan]])
+    @pytest.mark.parametrize("na_indices", [[1, 3]])
+    @pytest.mark.parametrize("na_position_first", ['first'])
+    @pytest.mark.parametrize("na_position_last", ['last'])
+    @pytest.mark.parametrize("column_name", ['c'])
```
Comment (WillAyd) [inline pandas/tests/frame/test_sorting.py]: Still not quite it as this will generate the Cartesian product of all of the items here.

Giving it a second look, this might be more complicated than I originally thought with parametrization to the point that it might overcomplicated things. I'd be OK with reverting to what you had before my request, and this could maybe be a separate PR if it even makes sense.

Sorry for the back and forth

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_sorting.py (hunk 8 above)
Comment (staftermath) [reply pandas/tests/frame/test_sorting.py]: no problem. I have reverted the parametrization.  Thanks for the careful checks.

**Comment by jreback** [comment]
this looked ok, @WillAyd ?

@staftermath can you merge master

**Review by WillAyd — APPROVED** [review]
lgtm; ok with merge after updating from master

**Comment by jreback** [comment]
thanks @staftermath
